File: Hashing/HashSet.py
class SimpleHashset:
    def __init__(self , size ):
        self.size = size 
        self.buckets = [[] for _ in range (size)]
    def hash_function(self , value):
        return sum(ord(char) for char in value ) % self.size
    def add(self,value):
        index = self.hash_function(value);
        bucket = self.buckets[index]
        if value not in bucket:
            bucket.append(value)
    def contains(self,value):
        index = self.hash_function(value)
        bucket = self.buckets[index]
        return value in bucket

    def remove(self,value):
        index = self.hash_function(value)
        bucket = self.buckets[index]
        if value in bucket:
            bucket.remove(value)
    def display(self):
        print("Hash Set")
        for index , bucket in enumerate(self.buckets):
            print(f"Bucket {index} : {bucket}")
if __name__ == '__main__':
    hash_set = SimpleHashset(10)
    names = ["Messi","Toll","Huyen","Drake"]
    for name in names :
        hash_set.add(name)
    hash_set.display()
    print("Toll co trong hashset khong " , hash_set.contains("Toll")) 
    print("Xoa mot value trong hashset")
    hash_set.remove("Drake")
    hash_set.display()
    print("Drake co trong hashset khong " , hash_set.contains("Drake"))

File: Hashing/HashSet.java
package DSA;

public class HashSet {
	int size;
	String[] bucket;

	public HashSet(int size) {
		this.size = size;
		this.bucket = new String[size];
	}

	public int hashFunction(String value) {
		int sumOfChar = 0;
		for (char c : value.toCharArray()) {
			sumOfChar += (int) c;
		}
		return sumOfChar % size;
	}

	public boolean contains(String value) {
		int index = hashFunction(value);
		if (bucket[index] == null)
			return false;
		if (bucket[index].equals(value))
			return true;
		return false;
	}

	public void add(String value) {
		int index = hashFunction(value);
		while (bucket[index] != null && !bucket[index].equals(value)) {
			index = (index + 1) % size;
		}
		bucket[index] = value;
	}

	public void remove(String value) {
		int index = hashFunction(value);
		if (bucket[index].equals(value))
			bucket[index] = null;
	}

	public void display() {
		for (int i = 0; i < size; i++) {
			System.out
					.println("Index " + i + ": " + (bucket[i] != null ? "\"" + "[" + bucket[i] + "]" + "\"" : "null"));
		}
	}

	public static void main(String[] args) {
		HashSet hs = new HashSet(10);
		System.out.println(" Danh sach ban dau ");
		hs.display();
		String[] myArray = { "Peter", "Messi", "Rose", "Toll", "Huyen" };
		for (String name : myArray) {
			hs.add(name);
		}
		System.out.println(" Danh sach sau khi them ");
		hs.display();
		String ele_remove = "Toll";
		System.out.println(" Danh sach sau khi xoa  " + ele_remove);
		hs.remove(ele_remove);
		hs.display();
		String ele_check = "Rose";
		System.out.print("Kiem tra " + ele_check + " co trong " + hs.contains(ele_check));
	}
}
